File: mobile/android/src/dev/tuist/hive/MobileClient.kt
package dev.tuist.hive

import android.util.Base64
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.security.SecureRandom

data class PreparedAuthorization(
    val url: String,
    val pending: PendingAuthorization,
)

data class RawResourceResult(
    val session: OAuthSession,
    val data: Any,
)

private data class NativeResponse(
    val status: Int,
    val body: String,
)

class MobileClient {
    companion object {
        const val REDIRECT_URI = "dev.tuist.hive://oauth2redirect"
    }

    private val transport = NativeHttpTransport()

    fun prepare(server: String): PreparedAuthorization {
        val effect = run(
            SharedCore.authorizationStart(
                server,
                REDIRECT_URI,
                randomSecret(),
                randomSecret(),
            ),
        )
        if (effect.requiredString("effect") != "browser") {
            throw MobileClientException("The shared core returned an invalid browser action.")
        }
        return PreparedAuthorization(
            url = effect.requiredString("authorization_url"),
            pending = PendingAuthorization(effect.requiredString("pending")),
        )
    }

    fun exchange(callbackUrl: String, pending: PendingAuthorization): OAuthSession {
        val effect = run(SharedCore.callbackStart(callbackUrl, pending))
        if (effect.requiredString("effect") != "session") {
            throw MobileClientException("The shared core returned an invalid session action.")
        }
        return OAuthSession(effect.requiredString("session"))
    }

    fun resource(session: OAuthSession, resource: HiveResource): RawResourceResult {
        val effect = run(SharedCore.resourceStart(session, resource, nowEpochSeconds()))
        if (effect.requiredString("effect") != "resource") {
            throw MobileClientException("The shared core returned an invalid resource action.")
        }
        return RawResourceResult(
            session = OAuthSession(effect.requiredString("session")),
            data = effect.get("data"),
        )
    }

    fun signOut(session: OAuthSession) {
        val effect = run(SharedCore.signOutStart(session))
        if (effect.requiredString("effect") != "done") {
            throw MobileClientException("The shared core returned an invalid sign-out action.")
        }
    }

    fun server(session: OAuthSession): String = SharedCore.server(session)

    private fun run(firstEffect: String): JSONObject {
        var effect = JSONObject(firstEffect)
        while (effect.requiredString("effect") == "http") {
            val response = transport.send(effect.getJSONObject("request"))
            effect = JSONObject(
                SharedCore.continueClient(
                    effect.requiredString("continuation"),
                    response.body,
                    response.status,
                    nowEpochSeconds(),
                ),
            )
        }
        return effect
    }

    private fun randomSecret(): String {
        val bytes = ByteArray(32)
        SecureRandom().nextBytes(bytes)
        return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
    }
}

private class NativeHttpTransport {
    fun send(plan: JSONObject): NativeResponse {
        val connection = URL(plan.requiredString("url")).openConnection() as HttpURLConnection
        try {
            connection.requestMethod = plan.requiredString("method")
            connection.connectTimeout = 10_000
            connection.readTimeout = 10_000
            connection.setRequestProperty("Accept", plan.requiredString("accept"))
            plan.optionalString("authorization")?.let {
                connection.setRequestProperty("Authorization", it)
            }
            plan.optionalString("body")?.let { body ->
                connection.doOutput = true
                connection.setRequestProperty(
                    "Content-Type",
                    plan.requiredString("content_type"),
                )
                connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            }

            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            return NativeResponse(
                status = status,
                body = stream?.bufferedReader()?.use { it.readText() }.orEmpty(),
            )
        } finally {
            connection.disconnect()
        }
    }
}

internal fun JSONObject.requiredString(key: String): String {
    val value = optString(key)
    if (value.isBlank()) {
        throw MobileClientException("Hive returned a response the application could not read.")
    }
    return value
}

internal fun JSONObject.optionalString(key: String): String? {
    if (isNull(key)) return null
    return optString(key).takeIf { it.isNotBlank() }
}

internal fun nowEpochSeconds(): Long = System.currentTimeMillis() / 1_000

class MobileClientException(message: String) : Exception(message)

File: mobile/ios/Hive/MobileClient.swift
import Foundation
import Security

struct PreparedAuthorization {
    let url: URL
    let pending: PendingAuthorization
}

struct ResourceResult<Value> {
    let session: OAuthSession
    let value: Value
}

private struct NativeRequestPlan: Decodable {
    let method: String
    let url: String
    let accept: String
    let body: String?
    let contentType: String?
    let authorization: String?

    enum CodingKeys: String, CodingKey {
        case method, url, accept, body, authorization
        case contentType = "content_type"
    }
}

private struct EffectHeader: Decodable {
    let effect: String
}

private struct HTTPEffect: Decodable {
    let effect: String
    let request: NativeRequestPlan
    let continuation: String
}

private struct BrowserEffect: Decodable {
    let effect: String
    let authorizationURL: String
    let pending: String

    enum CodingKeys: String, CodingKey {
        case effect, pending
        case authorizationURL = "authorization_url"
    }
}

private struct SessionEffect: Decodable {
    let effect: String
    let session: String
}

private struct ResourceEffect<Value: Decodable>: Decodable {
    let effect: String
    let session: String
    let data: Value
}

private struct NativeResponse {
    let status: Int
    let body: String
}

struct MobileClient {
    static let redirectURI = "dev.tuist.hive://oauth2redirect"

    private let core = SharedCore()
    private let transport = NativeHTTPTransport()

    func prepare(server: String) async throws -> PreparedAuthorization {
        let effect = try core.authorizationStart(
            server: server,
            redirectURI: Self.redirectURI,
            state: try randomSecret(),
            verifier: try randomSecret()
        )
        let final = try await run(effect)
        let browser: BrowserEffect = try decode(final)
        guard browser.effect == "browser", let url = URL(string: browser.authorizationURL) else {
            throw MobileClientError("The shared core returned an invalid browser action.")
        }
        return PreparedAuthorization(
            url: url,
            pending: PendingAuthorization(raw: browser.pending)
        )
    }

    func exchange(callbackURL: URL, pending: PendingAuthorization) async throws -> OAuthSession {
        let final = try await run(
            core.callbackStart(callbackURL: callbackURL.absoluteString, pending: pending)
        )
        let session: SessionEffect = try decode(final)
        guard session.effect == "session" else {
            throw MobileClientError("The shared core returned an invalid session action.")
        }
        return OAuthSession(raw: session.session)
    }

    func resource<Value: Decodable>(
        _ resource: HiveResource,
        session: OAuthSession,
        as _: Value.Type = Value.self
    ) async throws -> ResourceResult<Value> {
        let final = try await run(
            core.resourceStart(session: session, resource: resource, now: nowEpochSeconds())
        )
        let result: ResourceEffect<Value> = try decode(final)
        guard result.effect == "resource" else {
            throw MobileClientError("The shared core returned an invalid resource action.")
        }
        return ResourceResult(session: OAuthSession(raw: result.session), value: result.data)
    }

    func signOut(_ session: OAuthSession) async throws {
        let final = try await run(core.signOutStart(session: session))
        let result: EffectHeader = try decode(final)
        guard result.effect == "done" else {
            throw MobileClientError("The shared core returned an invalid sign-out action.")
        }
    }

    func server(_ session: OAuthSession) throws -> String {
        try core.server(session: session)
    }

    private func run(_ firstEffect: String) async throws -> String {
        var effect = firstEffect
        while true {
            let header: EffectHeader = try decode(effect)
            guard header.effect == "http" else { return effect }
            let request: HTTPEffect = try decode(effect)
            let response = try await transport.send(plan: request.request)
            effect = try core.continueClient(
                continuation: request.continuation,
                response: response.body,
                status: response.status,
                now: nowEpochSeconds()
            )
        }
    }

    private func randomSecret() throws -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        let status = bytes.withUnsafeMutableBytes { buffer in
            SecRandomCopyBytes(kSecRandomDefault, buffer.count, buffer.baseAddress!)
        }
        guard status == errSecSuccess else {
            throw MobileClientError("The application could not create a secure sign-in session.")
        }
        return Data(bytes)
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

private struct NativeHTTPTransport {
    private let session = URLSession.shared

    func send(plan: NativeRequestPlan) async throws -> NativeResponse {
        guard let url = URL(string: plan.url) else {
            throw MobileClientError("The shared core returned an invalid endpoint address.")
        }
        var request = URLRequest(url: url)
        request.httpMethod = plan.method
        request.setValue(plan.accept, forHTTPHeaderField: "Accept")
        if let authorization = plan.authorization {
            request.setValue(authorization, forHTTPHeaderField: "Authorization")
        }
        if let body = plan.body {
            request.httpBody = Data(body.utf8)
            request.setValue(plan.contentType, forHTTPHeaderField: "Content-Type")
        }

        let (data, response) = try await session.data(for: request)
        guard let response = response as? HTTPURLResponse,
              let body = String(data: data, encoding: .utf8)
        else {
            throw MobileClientError("Hive returned a response the application could not read.")
        }
        return NativeResponse(status: response.statusCode, body: body)
    }
}

private func decode<Value: Decodable>(_ value: String) throws -> Value {
    do {
        return try JSONDecoder().decode(Value.self, from: Data(value.utf8))
    } catch {
        throw MobileClientError("The shared core returned an invalid action.")
    }
}

private func nowEpochSeconds() -> Int64 {
    Int64(Date().timeIntervalSince1970)
}

struct MobileClientError: LocalizedError {
    let message: String

    init(_ message: String) {
        self.message = message
    }

    var errorDescription: String? { message }
}
